### ISSUE TITLE
Registry <-> web communication proactive security

### ISSUE THREAD

**Issue by dmp42** [docker/distribution#265]
People are going to communicate with their registry using webapps frontend / javascript (#39 is about enabling CORS to ease that) - as they will want to build nice hub-like UIs.

Also (although less likely) it might be possible for users to host random content on a registry, either by abusing it, or in the future, by extending it.

The bottom-line of this is that I would like us to take a defensive approach by default, and cram in additional headers into the registry so that we avoid (down the road) some class of web-based attacks on registries.

This one we should definitely consider:

1. `Strict-Transport-Security: max-age=16070400; includeSubDomains` help avoid TLS downgrading attacks - probably useful in all cases

These we might consider as well (by decreasing order of interest):

1. `X-Content-Type-Options: nosniff` - prevents IE from magically sniffing the content-type from the content (and force it to honor explicit content-type headers), thus protecting it from possibly rendering rogue content.
1. `X-Frame-Options: deny` - prevent embedding a registry url as an (i)frame in another page (would force a download / force rendering)
1. `Content-Security-Policy: default-src "self"` - if it is possible to host html on the registry, that would prevent third-party javascript, inline javascript, DOM-binded actions, and eval (http://content-security-policy.com/)
1. `X-XSS-Protection: 0` - if it's possible to host html on the registry, that would disable IE "XSS protection" which is more trouble than help (https://www.limessecurity.com/ie8xss/Abusing_IE8s_XSS_Filters.pdf)

**Comment by stevvooe** [comment]
Are there times where sending these headers would ever be undesired? For example, are there valid use cases where a user may want to turn them off?

**Comment by diogomonica** [comment]
HSTS does force you to have valid certs, even for internal websites. That might be painful for some.

In general, though, this is probably the behavior that we want: We ship this secure by default, but allow people to change it to their particular needs.

**Comment by stevvooe** [comment]
I'm wondering if we lock down the registry API aggressively, in favor of suggesting #206 as the main "web UI integration API". The registry API is really about access to resources, which is less pertinent in a browser. This would let us adopt these as a policy, avoiding the need to make this configurable at all.

One other consideration is to understand how API client implementations actually handle this. If our API implementations don't handle these headers, then what are they there for?

**Comment by ekristen** [comment]
Everything should be configurable. Having strict security measures in place by default is not a bad thing. Just make it so that people can switch them off, and make sure it is well documented.

The Registry API is the bread and butter, it is where all layer and image information is stored, downloaded and uploaded, locking it down by default should be aggressive IMO.

**Comment by aaronlehmann** [comment]
I think Strict-Transport-Security would be a problematic default, because people will often set up the registry with self-signed certificates before putting it in production. It would be annoying if the registry doesn't work at all without explicit steps to remove that header. I'd rather Strict-Transport-Security be something that the administrator opts into.

Also, `includeSubDomains` should be left out or made configurable, since it's very site-specific.

The rest seem reasonable, but I'm wondering if it's sufficient to just include `X-Content-Type-Options: nosniff`, and make sure we never serve any HTML content types (we don't, do we?). Then I believe the other options would be unnecessary, and it would keep things simpler to have a limited set of headers that we understand well.

There's a list of related HTTP headers here: https://www.owasp.org/index.php/List_of_useful_HTTP_headers. We might want to consider supporting public key pinning. This is another header that should be explicitly configured rather than being turned on out of the box.

In summary, I think we should throw in `X-Content-Type-Options: nosniff` on all HTTP responses, and introduce optional support for HSTS and possibly public key pinning. Does this sound good?

**Comment by stevvooe** [comment]
@aaronlehmann SGTM.

Perhaps, we should allow configuring static headers for inclusion to provide some control for admins.

**Comment by aaronlehmann** [comment]
Yes, I think it would be good to allow inclusion of static headers. Perhaps that could be a list of headers that defaults to `X-Content-Type-Options: nosniff`, and the admin could change the list as desired.

**Comment by stevvooe** [comment]
How about this:

```yaml
http:
  addr: :5000
  headers:
    X-Content-Type-Options: nosniff
```

This is what we have for the notification endpoints. Should be a fairly straightforward addition. The only issue with a default is that there needs to be a clear way to turn it off without sacrificing security. For example, if one adds a header, it should not remove `X-Content-Type-Options`. Simply treating empty headers as absent headers would be sufficient.

**Comment by aaronlehmann** [comment]
PR here: https://github.com/docker/distribution/pull/846

I decided not to build-in `X-Content-Type-Options: nosniff`, but instead to add it to the example configuration files. That means removing it simply involves deleting that line from the file, which is more intuitive than adding an empty header entry. It also provides a useful example in case users want to add their own headers.

**Comment by danleonard80** [comment]
Hi guys, So is it possible to enable HSTS for registry:2 ? Thanks a lot.
